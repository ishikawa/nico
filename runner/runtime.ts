import { StringDecoder } from "string_decoder";

export interface Printer {
  printlnNumber(n: number): number;
  printlnString(offset: number): number;
}

/**
 * Required `import` objects to run Nico compilation module.
 */
export interface WasmImportObject extends Record<string, Record<string, WebAssembly.ImportValue>> {
  "nico.runtime": { mem: WebAssembly.Memory };
  printer: {
    println_i32: Printer["printlnNumber"];
    println_str: Printer["printlnString"];
  };
}

export function buildImportObject(props: { memory: WebAssembly.Memory; printer: Printer }): WasmImportObject {
  const { memory, printer } = props;

  return {
    "nico.runtime": { mem: memory },
    printer: {
      println_i32: printer.printlnNumber.bind(printer),
      println_str: printer.printlnString.bind(printer)
    }
  };
}

export class ConsolePrinter implements Printer {
  stringView: StringView;

  constructor(memory: WebAssembly.Memory) {
    this.stringView = new StringView(memory);
  }

  printlnNumber(n: number): number {
    const string = n.toString();
    console.log(string);
    return string.length;
  }

  printlnString(offset: number): number {
    const string = this.stringView.getString(offset);
    console.log(string);
    return string.length;
  }
}

export class BufferedPrinter {
  stringView: StringView;
  buffer: string;

  constructor(memory: WebAssembly.Memory) {
    this.buffer = "";
    this.stringView = new StringView(memory);
  }

  printlnNumber(n: number): number {
    const string = n.toString();

    this.buffer += string;
    this.buffer += "\n";

    return string.length;
  }

  printlnString(offset: number): number {
    const string = this.stringView.getString(offset);

    this.buffer += string;
    this.buffer += "\n";

    return string.length;
  }
}

export class StringView {
  memory: WebAssembly.Memory;

  constructor(memory: WebAssembly.Memory) {
    this.memory = memory;
  }

  getString(base: number): string {
    const viewer = new DataView(this.memory.buffer, 0);
    const offset = viewer.getInt32(base, true);
    const length = viewer.getInt32(base + 4, true);

    const bytes = new Uint8Array(this.memory.buffer, offset, length);
    const decoder = new StringDecoder("utf-8");
    return decoder.end(Buffer.from(bytes));
  }
}
