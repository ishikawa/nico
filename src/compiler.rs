pub mod cli;
pub use cli::Command;

use crate::syntax::ParseError;
use std::io;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum CompilerError {
    #[error("invalid option: {0}")]
    InvalidOption(String),

    #[error(transparent)]
    InputSourceError(#[from] io::Error),

    #[error(transparent)]
    ParseError(#[from] ParseError),
}

impl From<String> for CompilerError {
    fn from(message: String) -> Self {
        CompilerError::InvalidOption(message)
    }
}
