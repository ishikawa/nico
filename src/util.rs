use std::cell::RefCell;
use std::rc::Rc;

pub fn wrap<T>(ty: T) -> Rc<RefCell<T>> {
    Rc::new(RefCell::new(ty))
}

pub trait UniqueNaming {
    fn next(&mut self) -> String;
}

/// Generates a unique string while continuously incrementing the index internally.
/// The user can specify a prefix for the generated string.
#[derive(Debug)]
pub struct SequenceNaming {
    type_var_index: usize,
    prefix: String,
}

impl SequenceNaming {
    pub fn new(prefix: &str) -> Self {
        Self {
            type_var_index: 0,
            prefix: prefix.to_string(),
        }
    }
}

impl UniqueNaming for SequenceNaming {
    fn next(&mut self) -> String {
        let i = self.type_var_index;
        self.type_var_index += 1;
        format!("{}{}", self.prefix, i)
    }
}
