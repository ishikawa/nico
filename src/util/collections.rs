use crate::arena::BumpaloVec;
use std::cell::Ref;

#[derive(Debug)]
pub struct RefVecIter<'r, 'arena, T> {
    vec: Ref<'r, BumpaloVec<'arena, T>>,
}

impl<'r, 'arena, T> RefVecIter<'r, 'arena, T> {
    pub fn new(vec: Ref<'r, BumpaloVec<'arena, T>>) -> Self {
        Self { vec }
    }
}

impl<'a: 'r, 'r, 'arena, T> IntoIterator for &'a RefVecIter<'r, 'arena, T> {
    type IntoIter = std::slice::Iter<'r, T>;
    type Item = &'r T;

    fn into_iter(self) -> Self::IntoIter {
        self.vec.iter()
    }
}
