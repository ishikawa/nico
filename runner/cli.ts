import fs from "fs";
import loadWabt from "wabt";
import { ConsolePrinter } from "./runtime";

function printUsage() {
  const usage = `
  Usage: npx ts-node cli program.wat
  `.trim();
  console.log(usage);
}

async function parseWat(filepath: string): Promise<Uint8Array> {
  const wabt = await loadWabt();

  return new Promise((resolve, _reject) => {
    const wasmModule = wabt.parseWat(filepath, fs.readFileSync(filepath, "utf8"));
    const { buffer } = wasmModule.toBinary({});
    resolve(buffer);
  });
}

export async function main(argv: string[]): Promise<number> {
  if (argv.length < 3) {
    printUsage();
    return 1;
  }

  const filepath = argv[2];
  const wasm = await parseWat(filepath);
  const module = await WebAssembly.compile(wasm);

  const memory = new WebAssembly.Memory({ initial: 1 });
  const printer = new ConsolePrinter(memory);

  const instance = await WebAssembly.instantiate(module, {
    js: { mem: memory },
    printer: {
      println_i32: printer.printlnNumber.bind(printer),
      println_str: printer.printlnString.bind(printer)
    }
  });

  const entryPoint = instance.exports.main;

  if (typeof entryPoint == "function") {
    entryPoint();
  }

  return 0;
}
