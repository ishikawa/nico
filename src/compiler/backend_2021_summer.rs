use crate::asm::wasm::Printer as WasmPrinter;
use crate::asm::AsmBuilder;
use crate::compiler::CompilerError;
use crate::parser::Parser;
use crate::syntax::Tokenizer;
use crate::CompilerPasses;

pub fn compile(src: String) -> Result<String, CompilerError> {
    let mut tokenizer = Tokenizer::from_string(&src);
    let parser = Parser::new();
    let mut module = parser.parse(&mut tokenizer)?;

    let mut passes = CompilerPasses::new();
    passes.apply(&mut module);

    let builder = AsmBuilder::new();
    let mut printer = WasmPrinter::new();

    let wasm_module = builder.build_module(&module);

    printer.pretty = true;
    printer.write_module(&wasm_module);
    Ok(printer.to_string())
}
