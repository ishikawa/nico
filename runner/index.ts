import { main } from "./cli";

main(process.argv).then(exitCode => {
  process.exitCode = exitCode;
});
