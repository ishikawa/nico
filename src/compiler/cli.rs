use clap::{App, Arg};
use io::Read;
use std::array;
use std::fmt;
use std::fs;
use std::io;
use std::iter::ExactSizeIterator;
use std::str::FromStr;

use super::backend_2021_spring;
use super::backend_2021_summer;
use super::CompilerError;

#[derive(Debug, Clone, Copy)]
pub enum CompilerBackend {
    Nico2021Spring,
    Nico2021Summer,
}

impl CompilerBackend {
    pub fn variants() -> impl Iterator<Item = CompilerBackend> {
        array::IntoIter::new([Self::Nico2021Spring, Self::Nico2021Summer])
    }
}

impl fmt::Display for CompilerBackend {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            CompilerBackend::Nico2021Spring => write!(f, "2021-spring"),
            CompilerBackend::Nico2021Summer => write!(f, "2021-summer"),
        }
    }
}

impl FromStr for CompilerBackend {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Self::variants()
            .find(|x| x.to_string() == s)
            .ok_or_else(|| format!("Unknown backend option: `{}`", s))
    }
}

#[derive(Debug)]
pub struct CompilerOptions {
    backend: CompilerBackend,
    filepath: Option<String>,
}

impl Default for CompilerOptions {
    fn default() -> Self {
        Self {
            backend: CompilerBackend::Nico2021Spring,
            filepath: None,
        }
    }
}

impl CompilerOptions {
    pub fn new() -> Self {
        Self::default()
    }
}

#[derive(Debug, Default)]
pub struct Command {}

impl Command {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn run(
        &self,
        args: impl ExactSizeIterator<Item = String>,
    ) -> Result<String, CompilerError> {
        let options = parse_options(args)?;

        let src = if let Some(filepath) = options.filepath {
            read_from_file(filepath.as_str())?
        } else {
            read_from_stdin()?
        };

        match options.backend {
            CompilerBackend::Nico2021Spring => backend_2021_spring::compile(src),
            CompilerBackend::Nico2021Summer => backend_2021_summer::compile(src),
        }
    }
}

fn parse_options(
    args: impl ExactSizeIterator<Item = String>,
) -> Result<CompilerOptions, CompilerError> {
    let mut options = CompilerOptions::new();

    // TODO: remove intermediate string object
    let backend_possible_values = CompilerBackend::variants()
        .map(|b| b.to_string())
        .collect::<Vec<_>>();
    let backend_possible_values: Vec<&str> =
        backend_possible_values.iter().map(AsRef::as_ref).collect();

    let matches = App::new("nico")
        .arg(
            Arg::with_name("backend")
                .long("backend")
                .takes_value(true)
                .possible_values(&backend_possible_values),
        )
        .arg(
            Arg::with_name("INPUT")
                .help("Sets the input file to use")
                .required(false)
                .index(1),
        )
        .get_matches_from(args);

    if let Some(backend) = matches.value_of("backend") {
        options.backend = backend.parse::<CompilerBackend>()?;
    }

    if let Some(filepath) = matches.value_of("INPUT") {
        options.filepath = Some(filepath.to_string());
    }

    Ok(options)
}

fn read_from_stdin() -> Result<String, io::Error> {
    let mut content = String::new();

    io::stdin().read_to_string(&mut content)?;

    Ok(content)
}

fn read_from_file(filename: &str) -> io::Result<String> {
    fs::read_to_string(filename)
}
