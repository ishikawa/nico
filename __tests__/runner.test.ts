import os from "os";
import fs from "fs";
import path from "path";
import { main } from "../runner/cli";
import { compileFileToWATFile } from "./util/compiler";

test("given empty input", async () => {
  const consoleSpy = jest.spyOn(console, "log").mockImplementation(() => {});

  let code = await main([]);

  expect(code).toEqual(1);
  expect(consoleSpy.mock.calls).toHaveLength(1);
  expect(consoleSpy.mock.calls[0][0]).toMatch(/Usage:/);

  consoleSpy.mockReset();
});

test("fizzbuzz", async () => {
  const consoleSpy = jest.spyOn(console, "log").mockImplementation(() => {});
  const outputFilepath = path.join(os.tmpdir(), "fizzbuzz.wat");

  await compileFileToWATFile("./samples/fizzbuzz.nico", outputFilepath);

  let code = await main(["node", "nico", outputFilepath]);

  expect(code).toEqual(0);
  expect(consoleSpy.mock.calls).toEqual([
    ["1"],
    ["2"],
    ["Fizz"],
    ["4"],
    ["Buzz"],
    ["Fizz"],
    ["7"],
    ["8"],
    ["Fizz"],
    ["Buzz"],
    ["11"],
    ["Fizz"],
    ["13"],
    ["14"],
    ["Fizz Buzz"]
  ]);

  consoleSpy.mockReset();
});
