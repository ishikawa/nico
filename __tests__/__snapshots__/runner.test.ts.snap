// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`./samples/fib.nico 1`] = `
Array [
  Array [
    "0",
  ],
  Array [
    "1",
  ],
  Array [
    "1",
  ],
  Array [
    "2",
  ],
  Array [
    "3",
  ],
  Array [
    "5",
  ],
  Array [
    "8",
  ],
  Array [
    "13",
  ],
  Array [
    "21",
  ],
  Array [
    "34",
  ],
  Array [
    "55",
  ],
  Array [
    "89",
  ],
  Array [
    "144",
  ],
  Array [
    "233",
  ],
  Array [
    "377",
  ],
  Array [
    "610",
  ],
  Array [
    "987",
  ],
  Array [
    "1597",
  ],
  Array [
    "2584",
  ],
  Array [
    "4181",
  ],
  Array [
    "6765",
  ],
]
`;

exports[`./samples/fizzbuzz.nico 1`] = `
Array [
  Array [
    "1",
  ],
  Array [
    "2",
  ],
  Array [
    "Fizz",
  ],
  Array [
    "4",
  ],
  Array [
    "Buzz",
  ],
  Array [
    "Fizz",
  ],
  Array [
    "7",
  ],
  Array [
    "8",
  ],
  Array [
    "Fizz",
  ],
  Array [
    "Buzz",
  ],
  Array [
    "11",
  ],
  Array [
    "Fizz",
  ],
  Array [
    "13",
  ],
  Array [
    "14",
  ],
  Array [
    "Fizz Buzz",
  ],
]
`;

exports[`./samples/max.nico 1`] = `
Array [
  Array [
    "-1",
  ],
  Array [
    "48",
  ],
  Array [
    "76",
  ],
  Array [
    "76",
  ],
  Array [
    "99",
  ],
  Array [
    "98",
  ],
]
`;

exports[`./samples/struct.nico 1`] = `
Array [
  Array [
    "1500",
  ],
]
`;

exports[`./samples/struct_pattern.nico 1`] = `
Array [
  Array [
    "7",
  ],
  Array [
    "8",
  ],
  Array [
    "99",
  ],
  Array [
    "88",
  ],
]
`;

exports[`./samples/sum.nico 1`] = `
Array [
  Array [
    "0",
  ],
  Array [
    "1666",
  ],
  Array [
    "-1215",
  ],
  Array [
    "2369",
  ],
]
`;
